File: Dynamic Programming/Fibonacci Series/FibonacciWithDPCode/TopDown/FibonacciWithDP_TD.java

public class Main
{
	public static void main(String[] args) 
    {
		int n = 8;
		Fib fib = new Fib(n);
        System.out.println(fib.getNthFibonacciNumber(n));
	}
}

class Fib
{
    private int[] _dp;
    
    public Fib(int n)
    {
        _dp = new int[n+1];
    }
    
    public int getNthFibonacciNumber(int n)
    {
        if(n == 1 || n == 0)
        {
            return n;
        }
        
        if(_dp[n] == 0)
        {
            _dp[n] = getNthFibonacciNumber(n-1) + getNthFibonacciNumber(n-2);
        }
        
        return _dp[n];
    }
}

File: Dynamic Programming/Fibonacci Series/FibonacciWithDPCode/TopDown/FibonacciWithDP_TD.ts
class FibWithDP_TD{
    _dp: number[];

    constructor(n: number) {
        this._dp = new Array(n+1);     
        this._dp[0] = 0;
        this._dp[1] = 1;   
    }

    getNthFibonacciNumber = (n: number): number => {
        if(n == 0 || n == 1){
            return n;
        }
        
        if(isNaN(this._dp[n])){
            this._dp[n] = this.getNthFibonacciNumber(n-1) + this.getNthFibonacciNumber(n-2);
        }
        return this._dp[n];
    }
}